libavcodec/libaribcaption.c

```diff
@@ -0,0 +1,1141 @@
+        ctx->renderer = NULL;
+    }
+    if (ctx->decoder) {
+        aribcc_decoder_free(ctx->decoder);
+        ctx->decoder = NULL;
+    }
+    if (ctx->context) {
+        aribcc_context_free(ctx->context);
+        ctx->context = NULL;
+     }
+
+    return 0;
+}
+
+static int aribcaption_init(AVCodecContext *avctx)
+{
+    ARIBCaptionContext *ctx = avctx->priv_data;
+    aribcc_profile_t profile;
+    aribcc_encoding_scheme_t encode_scheme;
+    int ret = 0;
+
+    ctx->avctx = avctx;
+
+    switch (avctx->profile) {
+    case FF_PROFILE_ARIB_PROFILE_A:
+        profile = ARIBCC_PROFILE_A;
+        /* assume 960x540 at initial state */
+        ctx->plane_width = 960;
+        ctx->plane_height = 540;
+        ctx->font_size = 36;
+        break;
+    case FF_PROFILE_ARIB_PROFILE_C:
+        profile = ARIBCC_PROFILE_C;
+        ctx->plane_width = 320;
+        ctx->plane_height = 180;
+        ctx->font_size = 16;
+        break;
+    default:
+        av_log(avctx, AV_LOG_ERROR, "Unknown or unsupported profile set.\n");
+        return AVERROR(EINVAL);
+    }
+    /* determine BorderStyle of ASS header */
+    if (ctx->ignore_background)
+        ctx->border_style = 1;
+    else
+        ctx->border_style = 4;
+    ctx->charstyle = ARIBCC_CHARSTYLE_DEFAULT;
+    if (ctx->force_stroke_text || ctx->ignore_background)
+        ctx->charstyle |= ARIBCC_CHARSTYLE_STROKE;
+
+    encode_scheme = ARIBCC_ENCODING_SCHEME_AUTO;
+    if (avctx->sub_charenc_mode == FF_SUB_CHARENC_MODE_DO_NOTHING)
+        encode_scheme = ARIBCC_ENCODING_SCHEME_ARIB_STD_B24_UTF8;
+
+    if (!(ctx->context = aribcc_context_alloc())) {
+        av_log(avctx, AV_LOG_ERROR, "Failed to alloc libaribcaption context.\n");
+        aribcaption_close(avctx);
+        return AVERROR_EXTERNAL;
+    }
+    aribcc_context_set_logcat_callback(ctx->context, logcat_callback, avctx);
+    if (!(ctx->decoder = aribcc_decoder_alloc(ctx->context))) {
+        av_log(avctx, AV_LOG_ERROR, "Failed to alloc libaribcaption decoder.\n");
+        aribcaption_close(avctx);
+        return AVERROR_EXTERNAL;
+    }
+    if (!aribcc_decoder_initialize(ctx->decoder,
+                                   encode_scheme,
+                                   ARIBCC_CAPTIONTYPE_CAPTION,
+                                   profile,
+                                   ARIBCC_LANGUAGEID_FIRST)) {
+        av_log(avctx, AV_LOG_ERROR, "Failed to initialize libaribcaption decoder.\n");
+        aribcaption_close(avctx);
+        return AVERROR_EXTERNAL;
+    }
+    aribcc_decoder_set_replace_msz_fullwidth_ascii(ctx->decoder,
+                                                   ctx->replace_fullwidth_ascii);
+
+    switch (ctx->subtitle_type) {
+    case SUBTITLE_ASS:
+        ret = set_ass_header(ctx);
+        if (ret != 0) {
+            av_log(avctx, AV_LOG_ERROR, "Failed to set ASS header: %s\n",
+                                        av_err2str(ret));
+            aribcaption_close(avctx);
```

Comment:
You have a lot of calls to `aribcaption_close` on error paths in this function. You can instead set `FF_CODEC_CAP_INIT_CLEANUP` in `caps_internal`, which will result in libavcodec automatically calling your `close` routine if initialization fails.

Reply:
Thanks. I'll try it.